File: code_snippets/ts/src/use_cases/async_tasks/fan_out_worker.ts
import * as restate from "@restatedev/restate-sdk/lambda";
import { Context, CombineablePromise } from "@restatedev/restate-sdk";

// <start_here>
const workerService = restate.service({
  name: "worker",
  handlers: {
    run: async (ctx: Context, task: Task) => {
      // Split the task in subtasks
      const subtasks: SubTask[] = await ctx.run("split task", () =>
        split(task)
      );

      const resultPromises = [];
      // <mark_1>
      for (const subtask of subtasks) {
        const subResultPromise = ctx
          .serviceClient(workerService)
          .runSubtask(subtask);
        // </mark_1>
        resultPromises.push(subResultPromise);
      }

      // <mark_2>
      const results = await CombineablePromise.all(resultPromises);
      // </mark_2>
      return aggregate(results);
    },

    runSubtask: async (ctx: Context, subtask: SubTask) => {
      // Processing logic goes here ...
      // Can be moved to a separate service to scale independently
    },
  },
});

// <mark_3>
export const handler = restate.endpoint().bind(workerService).handler();
// </mark_3>
// <end_here>

// ----------------------- Stubs to please the compiler -----------------------

type Task = {};
type Result = {};

type SubTask = {};
type SubTaskResult = void;

async function split(task: Task): Promise<SubTask[]> {
  return [];
}

async function aggregate(packages: SubTaskResult[]): Promise<Result> {
  return {};
}

File: code_snippets/python/src/use_cases/async_tasks/fan_out_worker.py
import restate
from pydantic import BaseModel
from restate import Service, Context


class Task(BaseModel):
    task_id: str


class Subtask(BaseModel):
    task_id: str
    subtask_id: str


def split(task) -> list[Subtask]:
    return list()


def aggregate(results) -> list[Task]:
    return list()


# <start_here>
worker_service = Service("worker")


@worker_service.handler()
async def run(ctx: Context, task: Task):
    subtasks = await ctx.run("split task", lambda: split(task))

    result_promises = []
    # <mark_1>
    for subtask in subtasks:
        sub_result_promise = ctx.service_call(run_subtask, arg=subtask)
        # </mark_1>
        result_promises.append(sub_result_promise)

    # <mark_2>
    results = [await promise for promise in result_promises]
    # </mark_2>
    return aggregate(results)


@worker_service.handler()
async def run_subtask(ctx: Context, subtask: Subtask):
    # Processing logic goes here...
    # Can be moved to a separate service to scale independently
    pass


# <mark_3>
app = restate.app([worker_service])
# </mark_3>
# <end_here>
